zerver/models.py

```diff
@@ -2481,22 +2482,25 @@ def is_history_public_to_subscribers(self) -> bool:
     ]
 
     @staticmethod
-    def get_client_data(query: QuerySet) -> List[Dict[str, Any]]:
+    def get_client_data(query: QuerySet) -> List[StreamClientDict]:
         query = query.only(*Stream.API_FIELDS)
         return [row.to_dict() for row in query]
 
-    def to_dict(self) -> Dict[str, Any]:
-        result = {}
-        for field_name in self.API_FIELDS:
-            if field_name == "id":
-                result["stream_id"] = self.id
-                continue
-            elif field_name == "date_created":
-                result["date_created"] = datetime_to_timestamp(self.date_created)
-                continue
-            result[field_name] = getattr(self, field_name)
-        result["is_announcement_only"] = self.stream_post_policy == Stream.STREAM_POST_POLICY_ADMINS
-        return result
+    def to_dict(self) -> StreamClientDict:
+        return StreamClientDict(
+            date_created=datetime_to_timestamp(self.date_created),
+            description=self.description,
+            first_message_id=self.first_message_id,
+            history_public_to_subscribers=self.history_public_to_subscribers,
+            stream_id=self.id,
+            invite_only=self.invite_only,
+            is_web_public=self.is_web_public,
+            message_retention_days=self.message_retention_days,
+            name=self.name,
+            rendered_description=self.rendered_description,
+            stream_post_policy=self.stream_post_policy,
+            is_announcement_only=self.stream_post_policy == Stream.STREAM_POST_POLICY_ADMINS,
+        )
```

Comment:
I guess one thing we lose with this change, in addition to the small bit of duplication, is the assertion that `API_FIELDS` matches this list; idunno if it's useful to assert that `API_FIELDS` matches the set of keys in this object, either here or in a test.

Reply:
I thought of using `StreamClientDict.__annotations.keys()` and replace it with `API_FIELDS` as a single source of truth, but `API_FIELDS` are not always fields of the dictionary so it requires some tweaking to remove some unwanted keys.
Another way will be adding test cases asserting the existence of `API_FIELDS` in these typed dictionaries.

Reply:
To be explicit, I'm imagining pseudocode like `assert set(API_FIELDS) == set(result.keys) + {"is_announcement_only"}` as a way to both ensure readability and also ensure if we mess up and add a field in only some of the places, we get an exception.

Reply:
I wrote a test case checking the `__annotations__` of the `TypedDict` types. Since the initialization of the dictionary is guarded by type annotations, as long as the code type checks it should have the expected fields.

Reply:
Cool.